libs/algo/odc/algo/_numexpr.py

```diff
@@ -0,0 +1,110 @@
+from typing import Dict, Optional, Any
+import dask
+import dask.array as da
+import numpy as np
+import xarray as xr
+import numexpr as ne
+import functools
+from ._dask import randomize, flatten_kv, unflatten_kv
+
+
+def apply_numexpr_np(
+    expr: str,
+    data: Optional[Dict[str, Any]] = None,
+    dtype=None,
+    out: Optional[np.ndarray] = None,
+    casting="safe",
+    order="K",
+    **params,
+) -> np.ndarray:
+    """
+    Apply numexpr to numpy arrays
+    """
+
+    if data is None:
+        data = params
+    else:
+        data.update(params)
+
+    if out is None and dtype is not None:
+        # This needs to be np.ndarray
+        arrays = [x for x in data.values() if isinstance(x, np.ndarray)]
+        if len(arrays) == 0:
+            raise ValueError("Could not find any arrays on input")
+
+        sample_input = arrays[0]
+        out = np.empty_like(sample_input, dtype=dtype)
+
+    return ne.evaluate(expr, local_dict=data, out=out, casting=casting, order=order)
+
+
+def apply_numexpr(
+    expr: str,
+    xx: xr.Dataset,
+    dtype=None,
+    name="result",
+    casting="safe",
+    order="K",
+    **params,
+):
+    """
+    Apply numexpr to variables within a Dataset
+    """
```

Comment:
As a public function I think it needs better docs.
- Dask friendly
- What are all the other args? Are they same as `xr.apply()` or `da.apply()`?
- Why would I want to use this? Is it just faster, or more RAM efficient too?